R/oauth-cache.R

```diff
@@ -0,0 +1,190 @@
+  }
+
+  if (!file.exists(cache)) {
+    create_cache(cache)
+  }
+  return(cache)
+}
+
+can_use_cache <- function(path = gargle_cache_path()) {
+  file.exists(path) || should_cache(path)
+}
+
+should_cache <- function(path = gargle_cache_path()) {
+  if (!interactive()) return(FALSE)
+
+  cat("Use a local file ('", path, "'), to cache OAuth access credentials ",
+      "between R sessions?\n", sep = "")
+  utils::menu(c("Yes", "No")) == 1
+}
+
+create_cache <- function(path = gargle_cache_path()) {
+
+  cache_parent <- dirname(path)
+  if (!dir.exists(cache_parent)) {
+    dir.create(cache_parent, recursive = TRUE)
+  }
+
+  file.create(path, showWarnings = FALSE)
+  if (!file.exists(path)) {
+    stop("Failed to create local cache ('", path, "')", call. = FALSE)
+  }
+
+  "!DEBUG cache exists: `path`"
+
+  # Protect cache as much as possible
+  Sys.chmod(path, "0600")
+
+  ## TODO(jennybc): this only looks in exact directory of cache path, not up
+  desc <- file.path(cache_parent, "DESCRIPTION")
+  if (file.exists(desc)) {
+    add_line(
+      file.path(cache_parent, ".Rbuildignore"),
+      paste0("^", gsub("\\.", "\\\\.", path), "$")
+    )
+  }
+  git <- file.path(cache_parent, c(".gitignore", ".git"))
+  if (any(file.exists(git))) {
+    add_line(
+      file.path(cache_parent, ".gitignore"),
+      path
+    )
+  }
+
+  TRUE
+}
+
+cache_token <- function(token, cache_path) {
+  "!DEBUG cache_token"
+  if (is.null(cache_path)) return()
+
+  tokens <- load_cache(cache_path)
+  tokens <- insert_token(tokens, token)
+  saveRDS(tokens, cache_path)
+}
+
+insert_token <- function(df, new) {
+  "!DEBUG insert_token"
+  if (length(df) == 0 || nrow(df) == 0) {
+    return(entibble(new))
+  }
+
+  df <- df[df$hash != new$hash(), ]
+  df[nrow(df) + 1, ] <- entibble(new)
+  df
+}
+
+entibble <- function(token) {
+  "!DEBUG entibble"
+  tibble::tibble(
+    hash = token$hash(),
+    token = list(token)
+  )
+}
+
+fetch_cached_token <- function(hash, cache_path) {
+  if (is.null(cache_path)) return()
+
+  tokens <- load_cache(cache_path)
+  tokens$token[[match(hash, tokens$hash)]]
+}
+
+
+fetch_matching_tokens <- function(hash, cache_path) {
+  if (is.null(cache_path)) return(NULL)
+
+  "!DEBUG `cache_path`"
+  tokens <- load_cache(cache_path)
+  matches <- mask_email(tokens$hash) == mask_email(hash)
+
+  if (!any(matches)) return(NULL)
+
+  tokens <- tokens[matches, ]
+
+  if (nrow(tokens) == 1) {
+    "!DEBUG Using a token cached for `extract_email(names(tokens))`"
+    return(tokens$token[[1]])
+  }
+
+  ## TODO(jennybc) if not interactive? just use first match? now I just give up
+  if (!interactive()) {
+    message("Multiple cached tokens exist. Unclear which to use.")
+    return(NULL)
+  }
+
+  emails <- extract_email(tokens$hash)
+  cat("Multiple cached tokens exist. Pick the one you want to use.\n")
+  cat("Or enter '0' to obtain a new token.")
+  this_one <- utils::menu(emails)
+
+  if (this_one == 0) return(NULL)
```

Comment:
And this too?

Reply:
This one can't `stop()` because it's the returned `NULL` that sets off the OAuth dance, i.e. it's how you can get a new token for an email, in the presence of existing tokens.